### PR TITLE
Add name() to C++ modules

### PR DESCRIPTION
**Pull request by goldsborough**
Gives C++ modules a `name`, useful for debugging and serialization

@ebetica @apaszke

### CODE REVIEW

File: torch/csrc/api/src/nn/module.cpp [hunk 1]
```diff
@@ -7,6 +7,13 @@
 #include <unordered_map>
 
 namespace torch { namespace nn {
+
+Module::Module(const char* name) : name_(name) {}
```
Comment (apaszke) [inline torch/csrc/api/src/nn/module.cpp]: Are we sure we always want to require people from specifying a name for the module? Maybe we could have a default implementation of `name()` that uses `typeid(*this).name()` (possibly with a roundtrip through our demangling code)?

**Comment by ezyang** [comment]
I agree with @paszke; having an RTTI based default seems like a good idea.

**Comment by goldsborough** [comment]
That will put the namespace in the name (`torch::nn::Linear`), but that may or may not be bad

**Comment by apaszke** [comment]
I know, but:
1. I'm ok with that
2. It's not that hard to get rid of it if we demangle the name.

**Comment by goldsborough** [comment]
Done

**Comment by ezyang** [comment]
```
19:47:07 /private/var/lib/jenkins/workspace/pytorch-builds/pytorch-macos-10.13-py3-build-test/test/cpp/api/container.cpp:39: FAILED:
19:47:07 due to a fatal error condition:
19:47:07   SIGSEGV - Segmentation violation signal
```

**Comment by goldsborough** [comment]
Actually turns out `this` always points to the base object inside constructors, so doing `typeid(*this)` in the constructor doesn't work. Instead I lazily initialize the name member the first time `name()` is called, since by then `this` will work polymorphically

**Review by ezyang — APPROVED** [review]

File: torch/csrc/api/src/nn/modules/dropout.cpp [hunk 2]
```diff
@@ -15,6 +20,10 @@ variable_list Dropout::forward(variable_list inputs) {
   return lst;
 }
 
+Dropout2d::Dropout2d(double p) : p_(p) {
+  assert(p < 1 && p >= 0);
```
Comment (apaszke) [inline torch/csrc/api/src/nn/modules/dropout.cpp]: Just FYI, asserts will usually compile to no-ops. We should use `TORCH_ASSERT` for checks

File: test/cpp/api/module.cpp [hunk 3]
```diff
@@ -36,6 +56,18 @@ TEST_CASE("module/zero-grad") {
   }
 }
 
+TEST_CASE("module/name") {
+  AGIUnit agi;
+  // Call it twice just to make sure there are no bugs in the lazy
+  // initialization semantics.
+  REQUIRE(agi.name() == "AGIUnit");
+  REQUIRE(agi.name() == "AGIUnit");
+  SECTION("correctly demangled") {
```
[the remark this answers is not in the document]
Comment (goldsborough) [reply test/cpp/api/module.cpp]: I can do something like that yes

File: torch/csrc/api/src/nn/modules/dropout.cpp (hunk 2 above)
Comment (goldsborough) [reply torch/csrc/api/src/nn/modules/dropout.cpp]: Yes will change that. Btw we should maybe canonicalize `AT_CHECK`  and `TORCH_ASSERT` ? I think only using `AT_CHECK` everywhere would make sense. @ezyang

**Pull request merged**